### PR TITLE
Add AWS EMR System tests

### PR DESCRIPTION
**Pull request by xinbinhuang**
Add system tests for two EMR related example DAGs.

The credentials are provided from `~/.aws/` following as in [boto3 configuration](https://boto3.amazonaws.com/v1/documentation/api/latest/guide/configuration.html#configuring-credentials). I feel like I should add some docs somewhere to indicate how to provide credentials for the system test, but not sure where... Please let me know if you would like me to add the documentation somewhere.

---
Make sure to mark the boxes below before creating PR: [x]

- [x] Description above provides context of the change
- [x] Unit tests coverage for changes (not needed for documentation changes)
- [x] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [x] Relevant documentation is updated including usage instructions.
- [x] I will engage committers as explained in [Contribution Workflow Example](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#contribution-workflow-example).

---
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).
Read the [Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines) for more information.

### CODE REVIEW

**Comment by mik-laj** [comment]
> I feel like I should add some docs somewhere to indicate how to provide credentials for the system test, but not sure where...

We have documentation about credentials in system tests:
https://github.com/apache/airflow/blob/master/TESTING.rst#forwarding-authentication-from-the-host. 
You can still add a small note in the docstring for the file: tests/providers/amazon/aws/operators/test_emr_system.py

The best solution is to add a marker that will skip the tests if the file with credentials is not available. If you try to run tests for GCP and you do not have credentials then you will see the message below:
```
========================================================================================= short test summary info ==========================================================================================
SKIPPED [1] /Users/kamilbregula/devel/google-airflow/airflow/tests/conftest.py:308: The test requires credential file /files/airflow-breeze-config/keys/gcp_life_sciences.json: <TestCaseFunction test_run_example_dag_function>
============================================================================================ 1 skipped in 1.06s ============================================================================================
```

**Comment by xinbinhuang** [comment]
> > I feel like I should add some docs somewhere to indicate how to provide credentials for the system test, but not sure where...
> 
> We have documentation about credentials in system tests:
> https://github.com/apache/airflow/blob/master/TESTING.rst#forwarding-authentication-from-the-host.
> You can still add a small note in the docstring for the file: tests/providers/amazon/aws/operators/test_emr_system.py
> 
> The best solution is to add a marker that will skip the tests if the file with credentials is not available. If you try to run tests for GCP and you do not have credentials then you will see the message below:
> 
> ```
> ========================================================================================= short test summary info ==========================================================================================
> SKIPPED [1] /Users/kamilbregula/devel/google-airflow/airflow/tests/conftest.py:308: The test requires credential file /files/airflow-breeze-config/keys/gcp_life_sciences.json: <TestCaseFunction test_run_example_dag_function>
> ============================================================================================ 1 skipped in 1.06s ============================================================================================
> ```

@mik-laj  Thanks for the suggestion. I took a look at the marker you mention, it needs to copy the credentials in the `files/breeze_config/keys` folder. For AWS system test, I think we can defer the discussion on sourcing credentials later as right now, we are still trying to add more helpers functions for AWS system test (i.e. #8581). So for this PR, I think it will better to use the default credentials for now, and we can change it after we have standardized the workflow.

What do you think?

**Comment by mik-laj** [comment]
@xinbinhuang I agree. We don't need it now, but you asked how to make the documentation. The best documentation is an intuitive program that has user-friendly messages

**Comment by xinbinhuang** [comment]
cc @feluelle @mustafagok - Feel free to take a look :)

@feluelle I can rebase on your changes once your PR is merged into master

File: tests/providers/amazon/aws/operators/test_emr_system.py
Comment (feluelle) [inline tests/providers/amazon/aws/operators/test_emr_system.py]: I think I would put this kind of code (`awscli`) into the aws_system_helpers.py file.

Comment (feluelle) [reply tests/providers/amazon/aws/operators/test_emr_system.py]: just encapsulate it in another function which you can then call here. (my opinion)

Comment (feluelle) [reply tests/providers/amazon/aws/operators/test_emr_system.py]: because sometimes one system test needs another system to fully test it?! Or if that is the case should we mock the other system? Do you understand what I mean? 😁

Comment (feluelle) [reply tests/providers/amazon/aws/operators/test_emr_system.py]: But I think the main reason why I would move it to the helper is because in case we want to switch the underlying program to a different one. For example from `awscli` to `boto3`?

**Review by feluelle — COMMENTED** [review]
Please add `emr.rst` docs in the `howto` section. Besides that LGTM. 

And if you PR is merged first I can rebase onto your changes (aws_system_helpers.py) - that's not a problem - you don't have to wait for my PR to be merged :)

File: tests/providers/amazon/aws/operators/test_emr_system.py
Comment (mustafagok) [inline tests/providers/amazon/aws/operators/test_emr_system.py]: naming convention suggestion, I have named "ECSSystemTest", after checking google system test class names.

```suggestion
class EmrSystemTest(SystemTest):
```

File: tests/providers/amazon/aws/operators/test_emr_system.py
Comment (mustafagok) [inline tests/providers/amazon/aws/operators/test_emr_system.py]: As @potiuk specified in slack, it should refer to backport packages.

Also, I think "aws_system_helpers.py" should be "amazon_system_helpers.py"

```suggestion
@pytest.mark.system("amazon")
```

File: tests/providers/amazon/aws/operators/test_emr_system.py
Comment (feluelle) [reply tests/providers/amazon/aws/operators/test_emr_system.py]: 👍

File: tests/providers/amazon/aws/operators/test_emr_system.py
Comment (feluelle) [reply tests/providers/amazon/aws/operators/test_emr_system.py]: suggestion LGTM 👍 

> Also, I think "aws_system_helpers.py" should be "amazon_system_helpers.py"

I did it like this because gcp is doing it the same way - `gcp_system_helpers.py`. But I am fine with `amazon_system_helpers.py`, too.

Comment (mik-laj) [reply tests/providers/amazon/aws/operators/test_emr_system.py]: We want to fix it. we are slowly updating each occurrence of the GCP word.
https://github.com/apache/airflow/pull/8697
https://github.com/apache/airflow/pull/8623
We are currently limiting major changes in system tests, because in this Q we want to run tests on CI.

**Comment by feluelle** [comment]
@xinbinhuang you can rebase onto latest master. The `amazon_system_helper.py` is there :)

File: tests/providers/amazon/aws/operators/test_emr_system.py
Comment (feluelle) [inline tests/providers/amazon/aws/operators/test_emr_system.py]:
```suggestion
```
You don't need that. It is automatically marked as amazon system test through deriving from `AmazonSystemTest`

File: tests/providers/amazon/aws/operators/test_emr_system.py
Comment (xinbinhuang) [reply tests/providers/amazon/aws/operators/test_emr_system.py]: I agree with you. I update it to move the function to the `AmazonSystemTest` class

**Comment by xinbinhuang** [comment]
@mustafagok @feluelle  Sorry for the late update, I have been swamped by other things in the last two weeks. I update the PR with your suggestions. PTAL :)

File: tests/providers/amazon/aws/operators/test_emr_system.py
Comment (xinbinhuang) [reply tests/providers/amazon/aws/operators/test_emr_system.py]: Nice. I didn't know that. Just fixed

File: tests/providers/amazon/aws/operators/test_emr_system.py
Comment (feluelle) [inline tests/providers/amazon/aws/operators/test_emr_system.py:27]: Shouldn't we remove these default roles at the end? So that you can call this tests over and over.
I also used [pytest fixtures](https://docs.pytest.org/en/latest/fixture.html).

Comment (xinbinhuang) [reply tests/providers/amazon/aws/operators/test_emr_system.py:27]: I don't think it's necessary to remove these default roles for running the tests. The command `aws emr create-default-roles` will only take effect if the associated roles for running EMR are not set in the AWS CLI config file. Otherwise, it will do nothing.

Comment (xinbinhuang) [reply tests/providers/amazon/aws/operators/test_emr_system.py:27]: To fully clean up the default roles created, it involves [six API calls](https://aws.amazon.com/premiumsupport/knowledge-center/emr-default-role-invalid/). I think in the future, there will be a dedicated managed AWS account for running system tests. So we can simply reuse the default roles.

Comment (xinbinhuang) [reply tests/providers/amazon/aws/operators/test_emr_system.py:27]: For pytest fixtures, it seems that all shared fixtures should go into the `contest.py` ( I cannot find this to be configurable at first glance. Please let me know if I am wrong). So it does not quite fit with how we store our fixtures at the moment (in the `*_system_helper.py` file and it is a mix of `unittest xunit-style setup` and `pytest fixtures` - maybe we can unify it in the future?). And I think the current code in this PR provides the least boilerplate code

Comment (feluelle) [reply tests/providers/amazon/aws/operators/test_emr_system.py:27]: Thanks for the explanation. Sounds good to me.

**Review by feluelle — APPROVED** [review]

**Comment by xinbinhuang** [comment]
@feluelle  @mustafagok  Thanks for reviewing and merging!

**Comment by potiuk** [comment]
Fantastics! Thanks @feluelle !

**Pull request merged**
